File: Kattis/cheatingstudents.hs
{-# LANGUAGE MultiWayIf #-}

import           Control.Applicative   (liftA2)
import           Control.Arrow         ((&&&), (>>>))
import           Control.Monad         (when)
import           Control.Monad.ST      (ST, runST)
import           Data.Array.Base       (listUArrayST, (!))
import           Data.Array.ST         (STUArray, index, newArray, readArray,
                                        writeArray)
import qualified Data.ByteString.Char8 as C
import           Data.Function         ((&))
import           Data.Functor          ((<&>))
import           Data.Ix               (Ix, rangeSize)
import           Data.List             (sort)
import           Data.Maybe            (fromJust)
import           Data.STRef            (STRef, modifySTRef', newSTRef,
                                        readSTRef)

main :: IO ()
main = do
    (n,_):xs <- C.getContents <&> (C.lines >>> map (C.words >>> map readInt >>> head &&& last))

    let ys = zip [0..n-1] xs    
            & edges
            & sort

    print $ runST $ do
        uf <- newUF (0,n-1)
        solve uf 0 ys

solve :: UF s Int -> Int -> [(Int,Int,Int)] -> ST s Int
solve uf ret [] = pure ret
solve uf ret ((c,u,v):xs) = do
        comps <- components uf
        conn <- connected uf u v
        if | comps == 1 -> pure ret
           | conn       -> solve uf ret xs
           | otherwise  -> merge uf u v >> solve uf (ret+c*2) xs

edges :: [(Int, (Int,Int))] -> [(Int,Int,Int)]
edges [] = []
edges ((i,(x,y)):xs) = map (\(j,(p,q)) -> (abs (x-p) + abs (y-q), i, j)) xs <> edges xs

data UF s i = UF {
    rng   :: (i,i),
    comps :: STRef    s Int,
    repr  :: STUArray s Int Int,
    sizes :: STUArray s Int Int
}

newUF :: Ix i => (i,i) -> ST s (UF s i)
newUF r = UF r <$> newSTRef n <*> listUArrayST (0,n-1) [0..] <*> newArray (0,n-1) 1
    where n = rangeSize r

find :: Ix i => UF s i -> i -> ST s Int
find uf u = find' (index (rng uf) u)
    where find' ix = do
            parent <- readArray (repr uf) ix
            if ix == parent
                then pure ix
                else do
                    grandparent <- readArray (repr uf) parent
                    writeArray (repr uf) ix grandparent
                    find' grandparent

merge :: Ix i => UF s i -> i -> i -> ST s Bool
merge uf u v = do
    p1 <- find uf u
    p2 <- find uf v
    if p1 == p2
        then pure False
        else do
            modifySTRef' (comps uf) pred
            s1 <- readArray (sizes uf) p1
            s2 <- readArray (sizes uf) p2
            if s1 < s2
                then do
                    writeArray (repr  uf) p1 p2
                    writeArray (sizes uf) p2 (s1 + s2)
                else do
                    writeArray (repr  uf) p2 p1
                    writeArray (sizes uf) p1 (s1 + s2)
            pure True

connected :: Ix i => UF s i -> i -> i -> ST s Bool
connected uf u v = liftA2 (==) (find uf u) (find uf v)

size :: Ix i => UF s i -> i -> ST s Int
size uf u = find uf u >>= readArray (sizes uf)

components :: UF s i -> ST s Int
components = comps >>> readSTRef

readInt :: C.ByteString -> Int
readInt = C.readInt >>> fromJust >>> fst

File: Kattis/cheatingstudents.cpp
#include <iostream>
#include <vector>
#include <iomanip>
#include <cstdint>
#include <algorithm>

using namespace std;
using i64 = int64_t;

template <typename Num>
class UF {
    vector<Num> repr;
    vector<Num> size;
    Num comps;

    public:
    UF(Num n){
        repr = vector<Num>(n, 0);
        size = vector<Num>(n, 1);
        comps = n;
        for (Num i {0}; i < n; repr[i++] = i);
    }
    Num find(Num u){
        if (repr[u] == u) return u;
        Num r = find(repr[u]);
        repr[u] = r;
        return r;
    }
    void merge(Num u, Num v){
        Num r1 = find(u);
        Num r2 = find(v);
        if (r1 == r2) return;
        Num s1 = size[r1];
        Num s2 = size[r2];
        if (s1 < s2){
            repr[r1] = r2;
            size[r2] += size[r1];
        }
        else{
            repr[r2] = r1;
            size[r1] += size[r2];
        }
        comps--;
    }
    Num size_of(Num u){ return size[find(u)]; }
    Num components(){ return comps; }
};

int main() {
    cin.tie(nullptr);
    ios::sync_with_stdio(false);
    cin.exceptions(ios::failbit);
    cout << setprecision(10) << fixed;

    i64 n, u, v;
    cin >> n;
    vector<pair<i64,i64>> xs(n);
    for (int i {0}; i < n; i++) {
        cin >> u >> v;
        xs[i] = {u,v};
    }
    vector<pair<i64,pair<int,int>>> edges;
    for (int i {0}; i < n; i++) {
        auto [x,y] = xs[i];
        for (int j {i+1}; j < n; j++) {
            auto [p,q] = xs[j];
            edges.push_back({abs(x-p) + abs(y-q), {i,j}});
        }
    }
    sort(edges.begin(), edges.end());

    UF uf = UF<int>(n);
    i64 ret = 0;

    for (auto [c, ij] : edges) {
        auto [i,j] = ij;
        if (uf.find(i) == uf.find(j)) continue;
        ret += c * 2;
        uf.merge(i,j);
        if (uf.components() <= 1) break;
    }

    cout << ret << '\n';
}
